### PR TITLE
[Slashing] store minted block in emanagerDB to avoid multiple block proposals in a round 

### PR DESCRIPTION
**Pull request by koseoyoung**
Works on #2375  , taking over PR #2436.
To cover edge cases(e.g., crash in the middle of the round and restart), make sure our protocol does not produce multiple block proposals in the same round. 

1. Add cachedMintedBlk in eManager/change the proto file.
2. Modify MintNewBlock() to use cached minted block.

### CODE REVIEW

File: consensus/scheme/rolldpos/endorsementmanager.go [hunk 1]
```diff
@@ -367,6 +380,21 @@ func (m *endorsementManager) AddVoteEndorsement(
...
+func (m *endorsementManager) MintBlock(blk *block.Block) error {
+	m.cachedMintedBlk = blk
+	if m.eManagerDB != nil {
+		return m.PutEndorsementManagerToDB()
+	}
+	return nil
+}
+
+func (m *endorsementManager) CachedMintedBlock() *block.Block {
+	if m.cachedMintedBlk != nil {
```
Comment (yutongp) [inline consensus/scheme/rolldpos/endorsementmanager.go:392]: just return `m.cachedMintedBlk`?

File: consensus/scheme/rolldpos/endorsementmanager.go [hunk 2]
```diff
@@ -375,6 +400,7 @@ func (m *endorsementManager) Cleanup(timestamp time.Time) error {
 	} else {
 		m.collections = map[string]*blockEndorsementCollection{}
 	}
+	m.cachedMintedBlk = nil
```
Comment (CoderZhi) [inline consensus/scheme/rolldpos/endorsementmanager.go:403]: every time, we "Prepare" a new round, the cachedMintedBlk will be clean up. So, this implementation may not work.

Comment (koseoyoung) [reply consensus/scheme/rolldpos/endorsementmanager.go:403]: Thanks for pointing it out! I fixed this issue by adding conditions when we clean up endorsementmanager. (Clean up only when the cached block timestamp is before round start time)

**Review by CoderZhi — COMMENTED** [review]
Finally, let's add some unit test to cover different cases

File: consensus/scheme/rolldpos/rolldposctx_test.go [hunk 3]
```diff
@@ -265,6 +260,85 @@ func TestCheckBlockProposer(t *testing.T) {
...
+
+	res, err := rctx.Proposal()
+	require.NoError(err)
+	ecm, ok := res.(*EndorsedConsensusMessage)
+	require.True(ok)
+	hash1, err := ecm.Document().Hash()
+	require.NoError(err)
+	height1 := ecm.Height()
+
+	res2, err := rctx.Proposal()
```
Comment (CoderZhi) [inline consensus/scheme/rolldpos/rolldposctx_test.go:331]: is it possible that the proposer has been switched between these two ".Proposal()" calls?

**Review by CoderZhi — APPROVED** [review]
The others look good, just one concern of the unit test

File: consensus/scheme/rolldpos/rolldposctx_test.go (hunk 3 above)
Comment (koseoyoung) [reply consensus/scheme/rolldpos/rolldposctx_test.go:331]: In this unit test, I assumed the case that the same proposer's node crashes and restart, so call Proposal() again. And make sure using cache from eManager.

Comment (koseoyoung) [reply consensus/scheme/rolldpos/rolldposctx_test.go:331]: So didn't consider proposer switching or round update

Comment (koseoyoung) [reply consensus/scheme/rolldpos/rolldposctx_test.go:331]: If the proposer has been switched in the middle, the changed proposer will not be the proposer of this round, so will skip it.

**Pull request merged**